src/field_theory/finite/chevalley_warning.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2019 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import data.mv_polynomial
+import algebra.pi_instances
+import field_theory.finite.basic
+
+/-!
+# The Chevalley–Warning theorem
+
+This file contains a proof of the Chevalley–Warning theorem.
+Throughout most of this file, K denotes a finite field with q elements.
+
+## Main results
+
+1. Let f be a multivariate polynomial in finitely many variables (X s, s : σ)
+   such that the total degree of f is less than (q-1) * cardinality of σ.
+   Then the evaluation of f on all points of `σ → K` (aka K^σ) sums to 0.
+   (`sum_mv_polynomial_eq_zero`)
+2. The Chevalley–Warning theorem (`char_dvd_card_solutions`).
+   Let (f i) be a finite family of multivariate polynomials
+   in finitely many variables (X s, s : σ) such that
+   the sum of the total degrees of the f i is less than the cardinality of σ.
+   Then the number of common solutions of the f i
+   is divisible by the characteristic of K.
+
+-/
+
+universes u v
+
+namespace finite_field
+open mv_polynomial function finset
+
+variables {K : Type*} [discrete_field K] [fintype K]
+variables {σ : Type*} [fintype σ] [decidable_eq σ]
+local notation `q` := fintype.card K
+
+lemma sum_mv_polynomial_eq_zero (f : mv_polynomial σ K)
+  (h : f.total_degree < (q - 1) * fintype.card σ) :
+  univ.sum (λ x, f.eval x) = (0:K) :=
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
+  rcases char_p.exists K with ⟨p, _char_p⟩, resetI,
+  rcases card K p with ⟨n, hp, hn⟩,
+  simp only [eval, eval₂, finsupp.sum, id.def],
+  rw [sum_comm, sum_eq_zero],
+  intros d hd,
+  rw [← mul_sum, mul_eq_zero], right,
+  simp only [finsupp.prod],
+  obtain ⟨i, hi⟩ : ∃ i, d i < q - 1,
+  { contrapose! h,
+    refine le_trans _ (finset.le_sup hd),
+    have : univ.sum (λ s:σ, q-1) ≤ univ.sum (λ s, d s) := sum_le_sum (λ s _, h s),
+    rw [sum_const, nat.smul_eq_mul, mul_comm, card_univ] at this,
+    rwa [finsupp.sum, show d.support = univ, from _],
```

Comment:
Like below, after `rw finsupp.sum, convert this using 1`, you have another application of `sum_extend_by_zero` or something like it. So you can ignore the suggested changes below, and again you don't actually need `hq`.

Reply:
Still a TODO, but Github thinks it's "outdated" because the code moved.

Reply:
Done.